[web-animations] support blending of mismatched filter lists

graouts opened the pull request:
#### e77c6ec7d3fe383255e32ab058d5844289fabbfb
<pre>
[web-animations] support blending of mismatched filter lists
<a href="https://bugs.webkit.org/show_bug.cgi?id=248270">https://bugs.webkit.org/show_bug.cgi?id=248270</a>

Reviewed by Simon Fraser.

We would fail to blend when filter lists mismatch and instead always fall back to the &quot;to&quot; value.

Instead of the matching filter functions on KeyframeEffect, we implement the specification for when
blending is supported in PropertyWrapperFilter::canInterpolate() such that we correctly identify
more discrete cases.

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/animation/filter-interpolation-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/animation/filter-interpolation-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-iteration-composite-operation-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFunc):
* Source/WebCore/animation/CSSPropertyBlendingClient.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setBlendingKeyframes):
(WebCore::KeyframeEffect::checkForMatchingFilterFunctionLists const): Deleted.
(WebCore::KeyframeEffect::checkForMatchingFilterFunctionLists): Deleted.
(WebCore::KeyframeEffect::checkForMatchingBackdropFilterFunctionLists): Deleted.
(WebCore::KeyframeEffect::checkForMatchingColorFilterFunctionLists): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:

Canonical link: <a href="https://commits.webkit.org/256975@main">https://commits.webkit.org/256975@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/051b01e9ab05ee274f3203bf190cb9025859c0bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6680 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30585 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106935 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167201 "Failed to checkout and rebase branch from PR 6758") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101384 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7002 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35429 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89824 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103609 "Failed to checkout and rebase branch from PR 6758") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103081 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84051 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/32256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/689 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/674 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5482 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2365 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1922 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/41213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1433,20 +1404,59 @@ class PropertyWrapperFilter final : public PropertyWrapper<const FilterOperation
+            return fromFilterOperations.isEmpty() != toFilterOperations.isEmpty();
+        };
+
+        auto listsMatchUpToIndex = [&](auto maxIndex) {
```
Might be clearer to say `numItems` instead of `maxIndex` to avoid any confusion about end-inclusiveness.

graouts replied:
Will change when landing.

smfr commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1433,20 +1404,59 @@ class PropertyWrapperFilter final : public PropertyWrapper<const FilterOperation
+            return fromFilterOperations.size() != toFilterOperations.size() && listsMatchUpToIndex(std::min(fromFilterOperations.size(), toFilterOperations.size()));
+        };
+
+        return compositeOperation != CompositeOperation::Replace || oneListIsEmpty() || listsMatchFully() || listsMatchPartially();
```
`listsMatchFully() || listsMatchPartially()` is potentially two list traversals, and if the lists are the same length (probably very common) they do the same thing. Maybe optimize for lists of the same length.

graouts replied:
Yes, I can definitely do that.

smfr commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-iteration-composite-operation-expected.txt:
```diff
@@ -15,8 +15,8 @@ PASS iteration composition of filter brightness for different unit animation
-FAIL iteration composition of discrete filter list because of mismatch of the order assert_equals: Animated filter list at 0s of the third iteration expected "brightness(1) contrast(1)" but got "contrast(2) brightness(2)"
-FAIL iteration composition of different length filter list animation assert_equals: Animated filter list at 50s of the first iteration expected "sepia(0.5) contrast(1.5)" but got "sepia(1) contrast(2)"
+PASS iteration composition of discrete filter list because of mismatch of the order
+FAIL iteration composition of different length filter list animation assert_equals: Animated filter list at 0s of the third iteration expected "sepia(1) contrast(3)" but got "sepia(1) contrast(4)"
```
Do you understand this failure?

graouts replied:
Yes, it will become a `PASS` when combined with https://github.com/WebKit/WebKit/pull/6762 at which points we'll pass each and every `filter` animation test in WPT.

graouts committed:
[web-animations] support blending of mismatched filter lists
https://bugs.webkit.org/show_bug.cgi?id=248270

Reviewed by Simon Fraser.

We would fail to blend when filter lists mismatch and instead always fall back to the "to" value.

Instead of the matching filter functions on KeyframeEffect, we implement the specification for when
blending is supported in PropertyWrapperFilter::canInterpolate() such that we correctly identify
more discrete cases.

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/animation/filter-interpolation-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/animation/filter-interpolation-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/keyframe-effects/effect-value-iteration-composite-operation-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFunc):
* Source/WebCore/animation/CSSPropertyBlendingClient.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setBlendingKeyframes):
(WebCore::KeyframeEffect::checkForMatchingFilterFunctionLists const): Deleted.
(WebCore::KeyframeEffect::checkForMatchingFilterFunctionLists): Deleted.
(WebCore::KeyframeEffect::checkForMatchingBackdropFilterFunctionLists): Deleted.
(WebCore::KeyframeEffect::checkForMatchingColorFilterFunctionLists): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:

Canonical link: https://commits.webkit.org/256975@main
